### ISSUE TITLE
[version3] Too many open files exception (OSX).

### ISSUE THREAD

**Issue by skaht** [libbitcoin/libbitcoin-server#350]
Any recommendations for the following:

**% git branch -v** (for libbitcoin-server)
```
* master 1d9edb8 Merge pull request #337 from evoskuil/master
```

**STDOUT**:
```
22:52:42.858416 INFO [blockchain] Block [406296] 1039 txs 2211 ins    0 wms  462 vms  209 vµs    4 rµs    5 cµs  131 pµs    4 aµs   64 sµs  108 dµs 0.000000
22:52:53.636974 INFO [network] Connected outbound channel [180.200.128.58:8333] (8)
22:53:29.140112 INFO [blockchain] Block [406297]  625 txs 1351 ins    0 wms  247 vms  183 vµs    5 rµs    5 cµs  123 pµs    2 aµs   48 sµs   88 dµs 0.000000
22:53:38.717243 INFO [blockchain] Block [406298] 2634 txs 4260 ins    0 wms 1193 vms  280 vµs    5 rµs    6 cµs  170 pµs    2 aµs   97 sµs   94 dµs 0.000000
22:53:42.008083 INFO [blockchain] Block [406299] 2420 txs 4593 ins    0 wms  996 vms  217 vµs    4 rµs    5 cµs  141 pµs    2 aµs   63 sµs   90 dµs 0.000000
22:53:44.186073 INFO [blockchain] Block [406300]  895 txs 4369 ins    0 wms  622 vms  142 vµs    4 rµs    4 cµs   68 pµs    2 aµs   64 sµs   61 dµs 0.000000
22:53:45.173052 INFO [blockchain] Block [406301] 1116 txs 2144 ins    0 wms  369 vms  172 vµs    5 rµs    7 cµs  102 pµs    3 aµs   55 sµs   80 dµs 0.000000
22:53:48.302806 INFO [blockchain] Block [406302] 1392 txs 3588 ins    0 wms  862 vms  240 vµs    4 rµs    5 cµs  117 pµs    3 aµs  112 sµs   77 dµs 0.000000
bs(64808,0x106290000) malloc: *** error for object 0x7fd009e1cd10: incorrect checksum for freed object - object was probably modified after being freed.
*** set a breakpoint in malloc_error_break to debug
```

**Comment by thecodefactory** [comment]
Thanks for reporting.  @skaht @evoskuil I'm not too familiar with crashes on OS X -- is it possible this could result in usage after a failed malloc rather than corruption via usage after free?

Also, I'm currently running (not the latest, but later than that): d01d8888fff2650da97f2c567c063edbc121b275

and did not see that issue, at least on GNU/Linux (debian).  Any chance you could try at least that revision, but perhaps latest?

**Comment by evoskuil** [comment]
If you are not running the tagged release bits on *all repos* you can certainly expect failures. There was at least one concurrency issue fixed in the last few commits. We've done extensive testing on Linux, OSX and Windows and haven't seen any such issue on release bits.

The error log is normal. The port 11050 is indicative of a [busted node (anarchistprime)](https://github.com/evoskuil/libbitcoin/blob/master/test/message/version.cpp#L27).

The debug log shows you seem to be running default configuration during initial block download. That's not generally recommended as it can be very slow and you may run your machine out of resources. You have 8 channels concurrently requesting potentially the same set of blocks, and you have each requesting tx relay. This can be very costly. The exception is not directly indicative of a memory or other resource constraint, but it can't be ruled out. But I'm more inclined to think this is a (likely fixed) concurrency bug.

In order to learn more you could run with valgrind, or at least collect a stack trace. But first you should rebuild all repos with release bits.

In any case, if the `flush_lock` file has been left behind your store is corrupt and must be discarded.

**Comment by evoskuil** [comment]
Ugh, accidentally closed.

**Comment by skaht** [comment]
Will do a new build shortly.  https://travis-ci.org/libbitcoin/libbitcoin-server looks okay for a static linking.

**Comment by evoskuil** [comment]
It's okay for dynamic as well :).

**Comment by skaht** [comment]
Static build completed, dynamic build failed with **--disable-static** set for the install-sh.

```
  CXX      test/test_libbitcoin_server_test-main.o
  CXX      test/test_libbitcoin_server_test-server.o
  CXXLD    test/libbitcoin_server_test
/Applications/Xcode.app/Contents/Developer/usr/bin/make  check-TESTS
FAIL: libbitcoin_server_test_runner.sh
===============================================
   libbitcoin-server 4.0.0: ./test-suite.log
===============================================

# TOTAL: 1
# PASS:  0
# SKIP:  0
# XFAIL: 0
# FAIL:  1
# XPASS: 0
# ERROR: 0

.. contents:: :depth: 2

FAIL: libbitcoin_server_test_runner.sh
======================================

dyld: Library not loaded: libboost_unit_test_framework.dylib
  Referenced from: /Users/Scott/Projects/BX/bitcoin-server/build/libbitcoin-server/test/.libs/libbitcoin_server_test
  Reason: image not found
./libbitcoin_server_test_runner.sh: line 21: 36742 Trace/BPT trap: 5       ./test/libbitcoin_server_test ${BOOST_UNIT_TEST_OPTIONS} > test.log
FAIL libbitcoin_server_test_runner.sh (exit status: 133)

============================================================================
Testsuite summary for libbitcoin-server 4.0.0
============================================================================
# TOTAL: 1
# PASS:  0
# SKIP:  0
# XFAIL: 0
# FAIL:  1
# XPASS: 0
# ERROR: 0
============================================================================
See ./test-suite.log
Please report to eric@voskuil.org
============================================================================
make[2]: *** [test-suite.log] Error 1
make[1]: *** [check-TESTS] Error 2
make: *** [check-am] Error 2
```

**Comment by evoskuil** [comment]
Did you use the installer?

**Comment by skaht** [comment]
Used https://github.com/libbitcoin/libbitcoin-server/blob/master/install.sh. Been some time since I did my own custom builds.

**Comment by skaht** [comment]
This latest build is kicking ass at the rate it building the Blockchain;-)

**Comment by evoskuil** [comment]
Please retry using https://github.com/libbitcoin/libbitcoin-server/blob/version3/install.sh and use the `--build-boost` and `--prefix` options (without sudo). This will isolate from your installed boost version.

**Comment by evoskuil** [comment]
Well it's fast, but not nearly as fast as it can be. Configuration makes a big difference still as well.

**Comment by skaht** [comment]
Computer appears to be running much cooler during the building of the mainnet Blockchain.

Will try a build using dynamically linked libraries using the **version3**  install.sh per your recommendation.

FYSA - Using the same config file that was associated withe the crash.  Diffed the custom config file against a production config file.  Differences were all accountable.

**Comment by evoskuil** [comment]
The default config is not ideal for initial block download. I believe I previously sent you some recommendations.

**Comment by skaht** [comment]
Besides changing the default location for the blockchain directory, *.log files, and the hosts.cache file, to  speed up the construction of  the local Blockchain image I've temporarily set:

flush_writes = false

**Comment by evoskuil** [comment]
See email.

**Comment by evoskuil** [comment]
Given the version question I'm closing. Please reopen if this surfaces in release code.

**Comment by skaht** [comment]
Last night's build crashed again when building the mainnet Blockchain.  Hence, here is precisely how I completed a static library build:

% wget -O bs-install.sh https://raw.githubusercontent.com/libbitcoin/libbitcoin-server/version3/install.sh
% chmod u+x bs-install.sh

% ./bs-install.sh **--disable-shared** --build-boost --prefix=/Users/Scott/Projects/BX/bitcoin-server/build/core --build-dir=/Users/Scott/Projects/BX/bitcoin-server/build

Tonight's static library link build completed just fine.  Afterwards, I renamed the build directory to build_static.  The following dynamically linked build failed at the tail end just like last night:

% ./bs-install.sh **--disable-static** --build-boost --prefix=/Users/Scott/Projects/BX/bitcoin-server/build/core --build-dir=/Users/Scott/Projects/BX/bitcoin-server/build

**Comment by skaht** [comment]
The libboost_unit_test_framework.dylib is not linking up but it was built.

% find build -name libboost_unit_test_framework.dylib -print
build/build-boost_1_57_0.tar.bz2/bin.v2/libs/test/build/clang-darwin-4.2.1/release/boost.locale.iconv-on/boost.locale.posix-on/threading-multi/libboost_unit_test_framework.dylib
build/core/lib/libboost_unit_test_framework.dylib

```
.
.
.
  CXXLD    test/libbitcoin_server_test
/Applications/Xcode.app/Contents/Developer/usr/bin/make  check-TESTS
FAIL: libbitcoin_server_test_runner.sh
===============================================
   libbitcoin-server 3.0.0: ./test-suite.log
===============================================

# TOTAL: 1
# PASS:  0
# SKIP:  0
# XFAIL: 0
# FAIL:  1
# XPASS: 0
# ERROR: 0

.. contents:: :depth: 2

FAIL: libbitcoin_server_test_runner.sh
======================================

dyld: Library not loaded: libboost_unit_test_framework.dylib
  Referenced from: /Users/Scott/Projects/BX/bitcoin-server/build/libbitcoin-server/test/.libs/libbitcoin_server_test
  Reason: image not found
./libbitcoin_server_test_runner.sh: line 21: 69244 Trace/BPT trap: 5       ./test/libbitcoin_server_test ${BOOST_UNIT_TEST_OPTIONS} > test.log
FAIL libbitcoin_server_test_runner.sh (exit status: 133)

============================================================================
Testsuite summary for libbitcoin-server 3.0.0
============================================================================
# TOTAL: 1
# PASS:  0
# SKIP:  0
# XFAIL: 0
# FAIL:  1
# XPASS: 0
# ERROR: 0
============================================================================
See ./test-suite.log
Please report to eric@voskuil.org
============================================================================
make[2]: *** [test-suite.log] Error 1
make[1]: *** [check-TESTS] Error 2
make: *** [check-am] Error 2
```

**Comment by evoskuil** [comment]
Do you mean to say that the build failed (in linkage) before completing tests, but you used the binary to sync mainnet anyway, and that process crashed while syncing.

**Comment by skaht** [comment]
Used the static library build to try to build the Blockchain.  Never tried to build the Blockchain with the dynamic library build.

**Comment by evoskuil** [comment]
I don't follow, are these two difference incidents then?

**Comment by skaht** [comment]
Yes they are.  I should have opened another ticket. 

Can't successfully create a dynamically linked bs executable on a Mac. Can successfully create a statically linked bs executable. A statically linked executable crashed again last night. Will try the new statically linked bs executable to see how far it can build the mainnet Blockchain.

If the building the Blockchain fails again, I'll give the production binary executable a test drive for building the mainnet Blockchain to see if the behaviors are the same.

**Comment by evoskuil** [comment]
Okay.

> A statically linked executable crashed again last night. Will try the new statically linked bs executable...

Does "*a* statically linked executable" refer to a good build of all release bits? If so what is meant by "the *new* statically linked bs executable"?

> I'll give the production binary executable a test drive...

Okay, by this we are talking about the release binary that is available for download.

**Comment by evoskuil** [comment]
> Can't successfully create a dynamically linked bs executable on a Mac.

The Travis OSX builds are taking a log time (currently timing out before completion) ever since the big outage. But as you can see from the older builds, this same configuration has linked and executed tests properly for a long time. However I'll run a dynamic build locally to ensure there is not a problem with the install itself.

**Comment by skaht** [comment]
Statically linked executable equates to a build using **--disable-shared**
Dynamically linked executable equates to a build using **--disable-shared**

The Blockchain tonight is building about x3 faster than last night:-)

**Comment by evoskuil** [comment]
Understood, but this refers to two static builds:

> A statically linked executable crashed again last night. Will try the new statically linked bs executable...

So I don't yet understand what is being differentiated.

**Comment by skaht** [comment]
Last night's build used install.sh #1 while tonight's build used #2:

1. https://raw.githubusercontent.com/libbitcoin/libbitcoin-server/master/install.sh
2. https://raw.githubusercontent.com/libbitcoin/libbitcoin-server/version3/install.sh

Currently running Darwin 14.5.0 Darwin Kernel Version 14.5.0.  Might bite the bullet an do the OSX upgrade soon.

**Comment by evoskuil** [comment]
Okay, thanks. There's nothing wrong with that OSX level.

> Last night's build crashed again...
> Last night's build used install.sh #1 while tonight's build used #2:

So your "last night" build used master bits from last night and you are presently running with version3 bits from tonight?

> The Blockchain tonight is building about x3 faster than last night:-)

The build bits should be identical, so I'm assuming you changed your configuration?

**Comment by evoskuil** [comment]
It looks like this line got dropped from some of our source code generation:

```xml
<option type="enable" name="shared" default="yes" define="BOOST_TEST_DYN_LINK" inherited="true" description="Required for dynamically linking boost test." />
```

Though I'm not sure why dynamic builds have been linking properly on Travis. It may be a difference when not using a prefix directory. In any case this should not be a problem apart from test execution.

**Comment by evoskuil** [comment]
See: https://github.com/libbitcoin/libbitcoin-server/issues/352

**Comment by skaht** [comment]
Do you want me to clean up issues now rolled up into #352 or just report to #352 when the issue is resolved from my perspective?

**Comment by evoskuil** [comment]
Sure, if you would remove the dyn link stuff from this issue that would be good.

**Comment by evoskuil** [comment]
Or better yet, just leave this closed and open a new issue for your malloc failure.

**Comment by skaht** [comment]
Crashed again.

```
09:53:32.110355 INFO [blockchain] Block [239168]  271 txs  704 ins    0 wms   13 vms   19 vµs   12 rµs    7 cµs    0 pµs    0 aµs    0 sµs  107 dµs 0.000000
09:53:32.171480 INFO [blockchain] Block [239169]  180 txs  368 ins    0 wms    4 vms   12 vµs    5 rµs    7 cµs    0 pµs    0 aµs    0 sµs   71 dµs 0.000000
09:53:32.376777 INFO [blockchain] Block [239170]  247 txs  523 ins    0 wms    7 vms   14 vµs    7 rµs    7 cµs    0 pµs    0 aµs    0 sµs   84 dµs 0.000000
09:53:32.468217 INFO [blockchain] Block [239171]  115 txs  321 ins    0 wms    4 vms   13 vµs    6 rµs    7 cµs    0 pµs    0 aµs    0 sµs   87 dµs 0.000000
libc++abi.dylib: terminating with uncaught exception of type boost::filesystem::filesystem_error: boost::filesystem::copy_file: Too many open files: "/Volumes/Blockchains/BS/debug.log", "/Users/Scott/Projects/BX/bitcoin-server/archive/debug-00019.log"
```
File /Users/Scott/Projects/BX/bitcoin-server/archive/debug-00019.log does not exist.

**Comment by evoskuil** [comment]
You hit system resource limits.

**Comment by evoskuil** [comment]
@skaht, do you have any information on the specific crash referenced here: https://github.com/libbitcoin/libbitcoin-server/issues/350#issuecomment-286592038

> Last night's build crashed again when building the mainnet Blockchain.

**Comment by evoskuil** [comment]
ping

**Comment by skaht** [comment]
[version4] build recreated the same issue.

```
12:44:56.794546 INFO [blockchain] Block [220999]  605 txs 1204 ins    0 wms   14 vms   12 vµs    6 rµs    6 cµs    0 pµs    0 aµs    0 sµs   82 dµs 0.000000
12:44:57.345935 INFO [blockchain] Block [221000]  336 txs  619 ins    0 wms    8 vms   14 vµs    7 rµs    7 cµs    0 pµs    0 aµs    0 sµs   69 dµs 0.000000
12:44:58.542531 INFO [blockchain] Block [221001]  707 txs 1116 ins    0 wms   14 vms   13 vµs    6 rµs    7 cµs    0 pµs    0 aµs    0 sµs   67 dµs 0.000000
libc++abi.dylib: terminating with uncaught exception of type boost::filesystem::filesystem_error: boost::filesystem::copy_file: Too many open files: "/Volumes/Blockchains/BS/debug.log", "12:44:56.794546 INFO [blockchain] Block [220999]  605 txs 1204 ins    0 wms   14 vms   12 vµs    6 rµs    6 cµs    0 pµs    0 aµs    0 sµs   82 dµs 0.000000
12:44:57.345935 INFO [blockchain] Block [221000]  336 txs  619 ins    0 wms    8 vms   14 vµs    7 rµs    7 cµs    0 pµs    0 aµs    0 sµs   69 dµs 0.000000
12:44:58.542531 INFO [blockchain] Block [221001]  707 txs 1116 ins    0 wms   14 vms   13 vµs    6 rµs    7 cµs    0 pµs    0 aµs    0 sµs   67 dµs 0.000000
libc++abi.dylib: terminating with uncaught exception of type boost::filesystem::filesystem_error: boost::filesystem::copy_file: Too many open files: "/Volumes/Blockchains/BS/debug.log", "/Users/Scott/Projects/BX/bitcoin-server/archive/debug-00045.log""
```

File **/Users/Scott/Projects/BX/bitcoin-server/archive/debug-00045.log**  is not found

What info do you want?

**Comment by evoskuil** [comment]
Thanks. I want to know if you have experienced any failures in a release build **other than** file handle excess.

**Comment by evoskuil** [comment]
Also, the file descriptor limit should be avoidable by reducing thread count. Try setting
```ini
[network]
threads = 1
```

**Comment by evoskuil** [comment]
I believe this will be mitigated by recent fixes to other issues in v3.1. It can never be eliminated because the implementation has no control over the platform's limits. However by reducing thread count it or raising platform limits it should always be manageable.

**Comment by evoskuil** [comment]
Given the resolution (currently in master) I'm closing this unless the problem recurs in that branch (to be pushed to v3.1).
